shiny/session/_session.py

```diff
@@ -1173,14 +1211,17 @@ def __init__(self, parent: Session, ns: ResolvedId) -> None:
         self._outbound_message_queues = parent._outbound_message_queues
         self._downloads = parent._downloads
 
+        self._root = parent.root_scope()
+        self.bookmark = BookmarkProxy(self)
+
     def _is_hidden(self, name: str) -> bool:
         return self._parent._is_hidden(name)
 
     def on_ended(
         self,
         fn: Callable[[], None] | Callable[[], Awaitable[None]],
     ) -> Callable[[], None]:
-        return self._parent.on_ended(fn)
+        return self._root.on_ended(fn)
```

Comment:
Try to do this with `_parent` instead of `_root`.

Reply:
Root has been removed. Everything is now through the parent